### ISSUE TITLE
[DevTools Bug]: Electron support broken in 4.27

### ISSUE THREAD

**Issue by Kilian** [facebook/react#25843]
### Website or app

n/a

### Repro steps

React devtools 4.27 no longer works in Electron[1] because `chrome.scripting` isn't implemented, which it started using due to the manifest v3 upgrade: https://github.com/facebook/react/pull/25145

When opening React devtools 4.27 in any Electron project (installed with [`electron-devtools-installer](https://github.com/MarshallOfSound/electron-devtools-installer/) for example) you'll see that the devtools panel is shown but no React components are ever found. Earlier versions worked well with Electron

[1] https://github.com/electron/electron/issues/36545 and https://github.com/MarshallOfSound/electron-devtools-installer/issues/232

### How often does this bug happen?

Every time

**Comment by mondaychen** [comment]
Ah I see. I'll try implement a fallback logic and release a fix.

**Comment by Kilian** [comment]
I wrote up some quick downgrade instructions for Electron apps (replace "Polypane" with your app name): https://polypane.app/docs/downgrading-react-devtools/

**Comment by serg06** [comment]
Thanks @Kilian ; I turned your fix into a script, tested on Windows in Git Bash

```bash
extension="data/extensions/fmkadmapgofadopljbjfkapdkoienihi"
manifest="$extension/manifest.json"

if ! [ -f $manifest ]; then
  echo manifest doesn\'t exist, cannot patch
  exit 0
fi

md5=`md5sum $manifest | awk '{split($0,a," "); print a[1]}'`
if [ "$md5" = "e821851e1ba9ff96244025afa8178b06" ]; then
  echo manifest is already correct
  exit 0
fi

echo fixing extension
curl https://polypane.app/fmkadmapgofadopljbjfkapdkoienihi.zip --output extensions-fix.zip
rm -r $extension
mkdir $extension
unzip extensions-fix.zip -d $extension
rm extensions-fix.zip
```

**Comment by filipefborba** [comment]
@mondaychen Hey, can I work on this issue? Any tips on how to start? Would be glad to help!

**Comment by mondaychen** [comment]
@filipefborba sorry, I think I put the "good first issue" label there by mistake. It's a bit to complex to set things up if you are new to this project.

**Comment by mondaychen** [comment]
Also my apologies to people who are impacted by this: I wanted to work on this earlier, but we are extremely understaffed at this moment, and I got pretty swamped in the past a few weeks. I should be able to work on this no later than next week.

**Comment by filipefborba** [comment]
@mondaychen no worries! If you find anything other "good first issue" or want/need some help, let me know. Also, I'm sorry that you've been swamped, that's never cool. Hope everything gets better! =)

**Comment by mondaychen** [comment]
tldr; This should be considered an Electron problem. For now, please follow the instructions at the end of this comment.

Hi everyone,

After some digging into the issue, I realized that it's not possible to create a version that works on both the latest Chrome & Electron, because:

- the old method of connecting to React requires an unsafe CSP, which is no longer allowed in Manifest V3
- the new method requires `chrome.scripting` to work, which is not available in Electron
- the fallback method does not work on Electron 
- It's not possible to dynamically provide a manifest.json file for an extension.

The only possible solution here is to revert the Manifest V3 upgrade. IMO that would be the wrong move. Chrome has announced that it will drop support for Manifest V2 by Jan 2023 (see https://developer.chrome.com/blog/mv2-transition/). Although they have decided to push back the hard deadline, it's unlikely that it will be pushed back forever. It's especially ironic that Electron also warns you "Manifest version 2 is deprecated and support will be removed in 2023" when you load an extension, while they don't have the API of Manifest V3.

So, in my opinion, Electron should add support for `chrome.scripting` as soon as possible. Before that, here is what I recommend:

**Workaround**
1. Download [this special Manifest V2 build of the extension](https://github.com/mondaychen/react/raw/017f120369d80a21c0e122106bd7ca1faa48b8ee/packages/react-devtools-extensions/ReactDevTools.zip)
1. Instead of replacing the existing extension in your Chrome data folder, unzip it to another location, because it's possible that our next release will overwrite it.
2. Update the path of `loadExtension`.

**Comment by madenney** [comment]
@eltaurito I just ran into the same problem with ERB as well

**Comment by xupea** [comment]
@eltaurito seems it's still using v3 of react devtool, and `chrome.scripting.registerContentScripts` is not supported by electron, can you try delete the Electron extension folder in your %APPDATA% , then `npm start`

**Comment by eltaurito** [comment]
> @eltaurito seems it's still using v3 of react devtool, and `chrome.scripting.registerContentScripts` is not supported by electron, can you try to delete the Electron extension folder in your %APPDATA% , then `npm start`

I deleted the extension folder, now it works perfectly 👍👍 👍 

Thank you a lot @xupea !

**Comment by kitschpatrol** [comment]
@xupea That worked for me as well, thanks so much.

On a Mac you can find Electron extensions here: `~/Library/Application Support/Electron/extensions`

**Comment by jonluca** [comment]
That version seems to be having issues with electron 24.

Also that repo redirects CRX requests to the developers domain, instead of downloading directly from the chrome webstore. (see here https://github.com/xupea/electron-devtools-installer/blob/master/src/downloadChromeExtension.ts#L24)

<img width="1062" alt="capture 2023-04-06 at 1 22 26 PM" src="https://user-images.githubusercontent.com/13029040/230486273-9450f04b-c448-469d-81f2-79bb79b99d90.png">


I published https://www.npmjs.com/package/electron-extension-installer to fix that (source code [here](https://github.com/jonluca/electron-extension-installer))


```typescript
import { installExtension, REACT_DEVELOPER_TOOLS } from "electron-extension-installer";

app.on("ready", async () => {
  await installExtension(REACT_DEVELOPER_TOOLS, {
    loadExtensionOptions: {
      allowFileAccess: true,
    },
  });
});
```

**Comment by xupea** [comment]
@jonluca what issues did you get?

**Comment by nekochan0122** [comment]
> That version seems to be having issues with electron 24.

I'm still getting errors in Electron 24, any ideas?

```
(node:17384) UnhandledPromiseRejectionWarning: TypeError: Function has non-object prototype 'undefined' in instanceof check
    at [Symbol.hasInstance] (<anonymous>)
    at Object.a3 (...)
    at Object.installExtension (...)
    at ...
(Use `electron --trace-warnings ...` to show where the warning was created)

(node:17384) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
```

![image](https://user-images.githubusercontent.com/47508893/235427189-b3b0cda2-669a-479e-9bec-3bf89beedeb5.png)

**Comment by yuwanlin** [comment]
![image](https://github.com/facebook/react/assets/18443683/e1d863ff-fd6b-4864-a779-d31094d0e7db)
@xupea, delete extensions still load failed.
```
import installer, { REACT_DEVELOPER_TOOLS, MOBX_DEVTOOLS } from 'electron-devtools-assembler';
// when ready
await installer(REACT_DEVELOPER_TOOLS, {
                loadExtensionOptions: {
                    allowFileAccess: true,
                },
            })
                .then((name) => this.logger.log(`Added Extension:  ${name}`))
                .catch((err) => this.logger.error('An error occurred: ', err));
```
electron version is 22，node version is 16.14.0

**Comment by Slapbox** [comment]
I have a similar problem with Electron 25.2.
![image](https://github.com/facebook/react/assets/6835891/1b338017-1f1e-40bf-9242-9e12acf272df)

**Comment by mulgurul** [comment]
Kilian, the script link does not exist anymore: "Page not found"
https://polypane.app/docs/downgrading-react-devtools/
BR Peter

**Comment by mulgurul** [comment]
Any news regarding this matter yet?

**Comment by Arkellys** [comment]
> Kilian, the script link does not exist anymore: "Page not found" https://polypane.app/docs/downgrading-react-devtools/ BR Peter

Using [Wayback Machine](https://web.archive.org/web/20221207185248/https://polypane.app/docs/downgrading-react-devtools/), edited with the [original instructions](https://github.com/facebook/react/issues/25843#issuecomment-1347907034) from author:

### 1. Download React-devtools 4.25
[Polypane] have prepared a zip file with react-devtools 4.25 here:
[fmkadmapgofadopljbjfkapdkoienihi.zip](https://web.archive.org/web/20221207185248/https://polypane.app/fmkadmapgofadopljbjfkapdkoienihi.zip)

### 2. Open [your app] extensions folder

Open the following folder in your explorer/finder:

- Windows: `%APPDATA%/[your app]/extensions/`
- macOS: `~/Library/Application Support/[your app]/extensions/`
- Linux: `~/.config/[your app]/extensions/`

### 3. Add or replace the current version with the version downloaded in step 1
If you have installed react-devtools:

- Find the folder `fmkadmapgofadopljbjfkapdkoienihi` and open it
- Remove all the files
- Extract the files from the zip file from step 1

If you do not have react-devtools installed:

- Create a folder `fmkadmapgofadopljbjfkapdkoienihi` and open it
- Extract the files from the zip file from step 1

### 4. Restart [your app]

After restarting [your app], react-devtools should work again. [Your app] will not automatically download new versions so unless you manually refresh, they should continue to work.

**Comment by k-villano** [comment]
hi im still having this problem using electron 27, I tried downloading the version of the extension using V2 but I'm a bit new and having trouble understanding the second and third steps of @mondaychen  suggestions.  I tried opening up the manifest json to see if that was the place to change it loadExtension or location, but it didn't seem right 

would appreciate any help, thanks

**Comment by Arkellys** [comment]
Should be fixed with Electron [v27.0.0](https://github.com/electron/electron/releases/tag/v27.0.0)+ (and also [v25.9.0](https://github.com/electron/electron/releases/tag/v25.9.0) and [v26.3.0](https://github.com/electron/electron/releases/tag/v26.3.0)) with [this pull request](https://github.com/electron/electron/pull/39395)

**Comment by Bartel-C8** [comment]
But DevTools 4.27.5+ has a bug with loading local file URLs in Electron:

https://github.com/facebook/react/issues/27749

**Comment by duddsA** [comment]
6127570289

**Comment by JordanDC2** [comment]
It appears `chrome.scripting` is available in Electron now. Can we update this project to use manifest version 3 now?

**Comment by goosewobbler** [comment]
@JordanDC2 I think it's [already using manifest v3](https://github.com/facebook/react/pull/25145)... there are other issues getting React Devtools working in Electron though:

https://github.com/facebook/react/issues/27749
https://github.com/electron/electron/issues/41613
https://github.com/electron/electron/issues/43092

**Comment by vitonsky** [comment]
Hey everyone. What's status on this bug?
I've update the electron to latest version 5 months ago and dev tools does not work anymore, so I can't profile our electron app.

**Comment by Nantris** [comment]
This is a really problematic issue; forced to fly blind or write custom measurements for performance improvements.

**Comment by vitonsky** [comment]
Hey @gaearon what about this problem?
As mentioned above by @Nantris - currently we are blind in terms of react performance debugging on electron platform.

That is critical problem for electron-only apps that cannot  be started in Chromium browser for debugging.
Is there plans to fix the problem with React dev tools in electron?

**Comment by Nantris** [comment]
> That is critical problem for electron-only apps that cannot be started in Chromium browser for debugging

Worth mentioning that due to Electron's security model, this is every real Electron app; every Electron app that isn't just wrapping a webpage and shipping a binary to do that for no good reason.
